File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/ProductEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.runtime.Quantity;

import java.math.BigDecimal;

/*

Product(argument List<Integer>) Integer
Product(argument List<Long>) Long
Product(argument List<Decimal>) Decimal
Product(argument List<Quantity>) Quantity

The Product operator returns the geometric product of the elements in source.

If the source contains no non-null elements, null is returned.

If the source is null, the result is null.

*/

public class ProductEvaluator {

    public static Object product(Object source) {
        if (source == null) {
            return null;
        }

        if (source instanceof Iterable) {
            Object result = null;
            for (Object element : (Iterable<?>) source) {
                if (element == null) return null;
                if (result == null) {
                    result = element;
                    continue;
                }
                if ((element instanceof Integer && result instanceof Integer)
                    || (element instanceof Long && result instanceof Long)
                    || (element instanceof BigDecimal && result instanceof BigDecimal)) {
                    result = MultiplyEvaluator.multiply(result, element);
                } else if (element instanceof Quantity && result instanceof Quantity) {
                    if (!((Quantity) element).getUnit().equals(((Quantity) result).getUnit())) {
                        // TODO: try to normalize units?
                        throw new IllegalArgumentException(
                                String.format(
                                        "Found different units during Quantity product evaluation: %s and %s",
                                        ((Quantity) element).getUnit(), ((Quantity) result).getUnit()
                                )
                        );
                    }
                    ((Quantity) result).setValue(
                            (BigDecimal) MultiplyEvaluator.multiply(
                                    ((Quantity) result).getValue(),
                                    ((Quantity) element).getValue()
                            )
                    );
                } else {
                    throw new InvalidOperatorArgument(
                        "Product(List<Integer>), Product(List<Long>), Product(List<Decimal>) or Product(List<Quantity>)",
                        String.format("Product(List<%s>)", element.getClass().getName())
                    );
                }
            }

            return result;
        }

        throw new InvalidOperatorArgument(
            "Product(List<Integer>), Product(List<Long>), Product(List<Decimal>) or Product(List<Quantity>)",
            String.format("Product(%s)", source.getClass().getName())
        );
    }

}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/elm/executing/ProductEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import java.math.BigDecimal
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Quantity

/*

Product(argument List<Integer>) Integer
Product(argument List<Long>) Long
Product(argument List<Decimal>) Decimal
Product(argument List<Quantity>) Quantity

The Product operator returns the geometric product of the elements in source.

If the source contains no non-null elements, null is returned.

If the source is null, the result is null.

*/
object ProductEvaluator {
    @JvmStatic
    fun product(source: Any?): Any? {
        if (source == null) {
            return null
        }

        if (source is Iterable<*>) {
            var result: Any? = null
            for (element in source) {
                if (element == null) return null
                if (result == null) {
                    result = element
                    continue
                }
                if (
                    (element is Int && result is Int) ||
                        (element is Long && result is Long) ||
                        (element is BigDecimal && result is BigDecimal)
                ) {
                    result = MultiplyEvaluator.multiply(result, element)
                } else if (element is Quantity && result is Quantity) {
                    require(element.unit == result.unit) {
                        String.format(
                            "Found different units during Quantity product evaluation: %s and %s",
                            element.unit,
                            result.unit,
                        )
                    }
                    result.value =
                        MultiplyEvaluator.multiply(result.value, element.value) as BigDecimal
                } else {
                    throw InvalidOperatorArgument(
                        "Product(List<Integer>), Product(List<Long>), Product(List<Decimal>) or Product(List<Quantity>)",
                        String.format("Product(List<%s>)", element.javaClass.name),
                    )
                }
            }

            return result
        }

        throw InvalidOperatorArgument(
            "Product(List<Integer>), Product(List<Long>), Product(List<Decimal>) or Product(List<Quantity>)",
            String.format("Product(%s)", source.javaClass.name),
        )
    }
}
